app/main/helpers/frameworks.py

```diff
@@ -365,7 +364,47 @@ def status_priority(status):
 
 
 def get_supplier_registered_name_from_declaration(declaration):
-    return(
+    return (
         declaration.get('supplierRegisteredName')  # G-Cloud 10 and later declaration key
         or declaration.get('nameOfOrganisation')  # G-Cloud 9 and earlier declaration key
     )
+
+
+class EnsureApplicationCompanyDetailsHaveBeenConfirmed:
+    """A decorator for framework application views that should not be accessible before company details have
+    been confirmed for that application.
+
+    This is a class-based decorator primarily so that it's possible to mock out the validator for the majority of tests
+    which aren't actively testing that the view requires suppliers to be in a certain application state."""
+
+    def __init__(self, data_api_client):
+        self.data_api_client = data_api_client
+
+    def __call__(self, func):
+        @wraps(func)
+        def decorated_view(*args, **kwargs):
+            if self.validator(*args, **kwargs):
+                return func(*args, **kwargs)
+
+            # Shouldn't be able to get here
+            abort(500, "There was a problem accessing this page of your application. Please try again later.")
+
+        return decorated_view
+
+    def validator(self, *args, **kwargs):
+        """Performs the actual validation that ensures the logged-in supplier has confirmed their company details
+        are correct for this application"""
+        if 'framework_slug' not in kwargs:
```

Comment:
Is it just luck that all the views we need to decorate take `framework_slug`? This couldn't be used on a view that didn't take it (if we needed to in future?)

Reply:
It's not luck, per se - the decorator is about locking them out of a part of their application to a framework, so it seems logical to me that we need to know _which_ framework they're applying to - so we should always have a reference to the framework slug.